### PR TITLE
Fancy formatting, converting Discord markdown <--> Pidgin HTML

### PR DESCRIPTION
**Pull request by bobbybee**
See commit messages.

### CODE REVIEW

**Comment by EionRobb** [comment]
Wow that's really nice and concise. It's also great that it doesn't need any 3rd party markdown lib (which I've used before in the rocket.chat and mattermost prpls)

The only comment I have, is how does it process URLs in markdown format?

**Comment by bobbybee** [comment]
AFAIK Discord's markdown is broken to the nth degree and doesn't even support links or images...

**Comment by bobbybee** [comment]
(This is not *quite* ready to be merged. I'm still trying to figure out why codeblocks only work half the time.)

File: libdiscord.c [hunk 1]
```diff
@@ -3312,20 +3369,55 @@ discord_replace_natural_emoji(const GMatchInfo *match, GString *result, gpointer
...
+
+	return html;
+}
+
+static gchar*
+discord_escape_md(gchar* markdown)
+{
+	markdown = discord_helper_replace(markdown, "_", "\\_");
+	markdown = discord_helper_replace(markdown, "*", "\\*");
+	markdown = discord_helper_replace(markdown, "~", "\\~");
```
Comment (EionRobb) [inline libdiscord.c]: Do `\`s need to be escaped too?

File: libdiscord.c [hunk 2]
```diff
@@ -1412,6 +1412,61 @@ discord_replace_emoji(const GMatchInfo *match, GString *result, gpointer user_da
 	return FALSE;
 }
 
+#define HTML_TOGGLE_OUT(f, a, b) out = g_string_append(out, f ? b : a); f = !f;
+
+static gchar*
+discord_convert_markdown(gchar* html) {
```
Comment (EionRobb) [inline libdiscord.c]: `gchar *html` should be changed to `const gchar *html` to signify that `html` is not going to be modified (and for the compiler to help catch when it accidentally is modified)

File: libdiscord.c [hunk 3]
```diff
@@ -1488,6 +1543,8 @@ discord_process_message(DiscordAccount *da, JsonObject *data)
 		escaped_content = tmp;
 	}
 
+	escaped_content = discord_convert_markdown(escaped_content);
```
Comment (EionRobb) [inline libdiscord.c]: This leaks the original `escaped_content`

**Comment by bobbybee** [comment]
OK, I'm satisfied with this. The earlier bugs were solved over IRC.

(Hint: I don't know how to use Pidgin correctly.)

File: libdiscord.c [hunk 4]
```diff
@@ -1412,6 +1412,66 @@ discord_replace_emoji(const GMatchInfo *match, GString *result, gpointer user_da
...
+			if(html[i + 1] == '_') {
+				HTML_TOGGLE_OUT(s_underline, "<u>", "</u>");
+			} else {
+				HTML_TOGGLE_OUT(s_italics, "<i>", "</i>");
+			}
+		} else if(c == '`') {
+			if(html[i + 1] == '`' && html[i + 2] == '`') {
+				if(!s_codeblock) {
+					while(html[i] != '\n' && html[i] != ' ' && html[i]) ++i;
+					out = g_string_append(out, "<br/><span style='font-family: monospace'>");
```
Comment (EionRobb) [inline libdiscord.c]: Does this need a `white-space: pre` in the style to allow for multiple spaces to not be collapsed?

**Comment by bobbybee** [comment]
Bug discovered: o_o renders incorrectly. Investigating...

**Comment by bobbybee** [comment]
Solving the peculiarities of Discord's specific implementation is more complex than I anticipated. Still working on "o_o" with some friends who helped uncover a number of related bugs.

**Comment by bobbybee** [comment]
There are still a handful of subtle differences between this parser and Discord's when parsing non-compliant markdown, but this should be usable in normal conversation. I'll continue to test my bleeding edge version to make sure I didn't miss anything important.

**Comment by EionRobb** [comment]
Yay, pretty formatting \o/

**Comment by EionRobb** [comment]
Just needed a tiny tweak for Windows

**Pull request merged**